certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+rule integrityOfCollect() {
+    env e; uint256 userId; uint256 assetId;
+    uint128 collectedAmt; uint128 collectableBefore; uint128 collectableAfter;
+
+    collectableBefore = collectable(e, userId, assetId);
+    collectedAmt = collect(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert collectableAfter == 0;
+}
+
+
+/// @notice Calling the method collect() should never revert
+rule revertCharacteristicsOfCollect() {
+    env e; uint256 userId; uint256 assetId; uint128 collectedAmt;
+
+    require e.msg.sender != 0;  // safe assumption that prevents revert
+    require e.msg.value == 0;  // safe assumption that prevents revert
+
+    collectedAmt = collect@withrevert(e, userId, assetId);
+    assert !lastReverted;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// @param amt The given amount
+/// After give of amt, the splittable of the receiver should increase exactly by amt
+rule correctnessOfGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint128 splittableBefore; uint128 splittableAfter;
+
+    splittableBefore = splittable(e, receiver, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, receiver, assetId);
+
+    assert splittableAfter == splittableBefore + amt;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// The splittable of any other user that is not receiver should not change
+rule splittableOfNonReceiverNotAffectedByGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint256 otherUser; uint128 splittableBefore; uint128 splittableAfter;
+
+    require otherUser != receiver;
+
+    splittableBefore = splittable(e, otherUser, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, otherUser, assetId);
+
+    assert splittableAfter == splittableBefore;
+}
+
+
+/// @notice the method hashSplits() Calculates the hash of the list of splits receivers.
+/// We use boolean selector to operate on two different lists of splits receivers.
+/// We verify that two calculated hashes are the same only if they got exactly the same input
+/// We also verify that different inputs must generate different hashes
+rule correctnessOfHashSplits() {
+    env e; uint256 index; uint256 length1; uint256 length2;  
+    uint256 userId1; uint32 weight1; uint256 userId2; uint32 weight2;
+    bytes32 receiversHash1; bytes32 receiversHash2;
+
+    length1 = getCurrSplitsReceiverLocaLength(e, true);
+    length2 = getCurrSplitsReceiverLocaLength(e, false);
+
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index);
+    userId2, weight2 = getCurrSplitsReceiverLocalArr(e, false, index);
+
+    receiversHash1 = hashSplits(e, true);
+    receiversHash2 = hashSplits(e, false);
+
+    assert (receiversHash1 == receiversHash2) => (length1 == length2);
```

Comment:
Unless there is a time-out issue, these three asserts could be combined to one.

Reply:
OK, modified the code.